Olá, Iago

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.

Como pontos fortes do projeto, destaco:
- Código simples e eficiente (sem loop desnecessário)
- Uso adequado dos recursos da lição
- Calibrou hiperparâmetros, usou corretamente treino,validação e teste
      
Bom trabalho. Porém, faltou avaliar os moelos ajustados no conjunto de teste.


</div>




<div class="alert alert-block alert-success">
<b> Comentário Geral do revisor: </b> <a class="tocSkip"></a>

Tudo certo agora. Parabéns pelo empenho. Desejo sucesso na jornada.

</div>

In [1]:
# Importação das bibliotecas essenciais

import pandas as pd          # manipulação e análise de dados
import numpy as np           # operações numéricas
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [2]:
# Leitura do arquivo CSV e visualização inicial

df = pd.read_csv('/datasets/users_behavior.csv')

# Exibição das primeiras linhas do DataFrame
df.head()

,calls,minutes,messages,mb_used,is_ultra
0,40.0,311.90,83.0,19915.42,0
1,85.0,516.75,56.0,22696.96,0
2,77.0,467.66,86.0,21060.45,0
3,106.0,745.53,81.0,8437.39,1
4,66.0,418.74,1.0,14502.75,0


### Leitura e visualização inicial dos dados

Nesta etapa, foi realizada a **importação do arquivo `users_behavior.csv`** localizado no caminho `datasets/`.
O objetivo foi confirmar que o arquivo foi lido corretamente e obter uma **visão inicial da estrutura dos dados**.

O DataFrame contém **cinco colunas**:

| Coluna     | Descrição                                                |
| ---------- | -------------------------------------------------------- |
| `calls`    | Número de chamadas realizadas pelo cliente durante o mês |
| `minutes`  | Duração total das chamadas em minutos                    |
| `messages` | Quantidade de mensagens de texto enviadas                |
| `mb_used`  | Quantidade de dados de internet usados (em MB)           |
| `is_ultra` | Plano utilizado (0 = Smart, 1 = Ultra)                   |

A visualização inicial mostra que os dados estão **limpos e bem estruturados**, sem valores ausentes aparentes, e as colunas estão em formatos adequados para análise.
A próxima etapa será realizar uma **análise exploratória básica**, verificando o tamanho do dataset, tipos de dados e estatísticas descritivas.

In [3]:
# Análise exploratória inicial do dataset

# Dimensões do DataFrame
print('Dimensões do dataset (linhas, colunas):', df.shape)

# Informações gerais sobre os tipos de dados
print('\nInformações sobre colunas e tipos de dados:')
df.info()

# Verificação de valores ausentes
print('\nValores ausentes por coluna:')
print(df.isna().sum())

# Estatísticas descritivas básicas
print('\nEstatísticas descritivas das variáveis numéricas:')
display(df.describe())

# Distribuição da variável-alvo
print('\nDistribuição da variável-alvo (is_ultra):')
print(df['is_ultra'].value_counts())
print('\nProporção percentual:')
print(df['is_ultra'].value_counts(normalize=True) * 100)

Dimensões do dataset (linhas, colunas): (3214, 5)

Informações sobre colunas e tipos de dados:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB

Valores ausentes por coluna:
calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

Estatísticas descritivas das variáveis numéricas:


,calls,minutes,messages,mb_used,is_ultra
count,3214.000000,3214.000000,3214.000000,3214.000000,3214.000000
mean,63.038892,438.208787,38.281269,17207.673836,0.306472
std,33.236368,234.569872,36.148326,7570.968246,0.461100
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,40.000000,274.575000,9.000000,12491.902500,0.000000
50%,62.000000,430.600000,30.000000,16943.235000,0.000000
75%,82.000000,571.927500,57.000000,21424.700000,1.000000
max,244.000000,1632.060000,224.000000,49745.730000,1.000000



Distribuição da variável-alvo (is_ultra):
0    2229
1     985
Name: is_ultra, dtype: int64

Proporção percentual:
0    69.352831
1    30.647169
Name: is_ultra, dtype: float64


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Está correto. Senti falta de gráficos

</div>


### Análise exploratória inicial

Após examinar o dataset `users_behavior.csv`, pode-se observar as seguintes características:

* O conjunto de dados contém **3.214 linhas** e **5 colunas**.
* Todas as colunas apresentam **valores não nulos**, ou seja, **não há dados ausentes**.
* Os tipos de dados estão apropriados:

  * `calls`, `minutes`, `messages`, `mb_used`: `float64` (valores numéricos contínuos);
  * `is_ultra`: `int64` (variável categórica binária para classificação).

#### Estatísticas descritivas

* O número médio de chamadas é de **63 por mês**, com duração média de **428 minutos**.
* O uso médio de internet é de aproximadamente **17 GB (17.000 MB)**.
* O número médio de mensagens enviadas é de **38 por mês**.
* As variáveis exibem variação considerável (desvios padrão altos), o que indica **diferenças significativas de comportamento entre os clientes**.

#### Variável-alvo (`is_ultra`)

* Plano **Smart (0)**: 2.229 clientes (~69%)
* Plano **Ultra (1)**: 985 clientes (~31%)

In [4]:
# Divisão dos dados em conjuntos de treino, validação e teste

# Separando as features (X) e o alvo (y)
X = df.drop('is_ultra', axis=1)
y = df['is_ultra']

# Primeiro, dividimos entre treino+validação e teste
X_temp, X_test, y_temp, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Em seguida, dividimos o conjunto temporário em treino e validação
X_train, X_valid, y_train, y_valid = train_test_split(
    X_temp, y_temp, test_size=0.25, random_state=42, stratify=y_temp
)
# (0.25 de 0.8 = 0.2 → total: 60% treino, 20% validação, 20% teste)

# Conferindo tamanhos finais
print('Tamanho do conjunto de treino:', X_train.shape[0])
print('Tamanho do conjunto de validação:', X_valid.shape[0])
print('Tamanho do conjunto de teste:', X_test.shape[0])

Tamanho do conjunto de treino: 1928
Tamanho do conjunto de validação: 643
Tamanho do conjunto de teste: 643


### Markdown — Etapa 4: Divisão dos dados em treino, validação e teste

Para garantir a avaliação justa dos modelos, o conjunto de dados foi dividido em três partes:

| Conjunto      | Percentual | Tamanho (linhas) | Finalidade                                             |
| ------------- | ---------- | ---------------- | ------------------------------------------------------ |
| **Treino**    | 60%        | 1.928            | Usado para treinar os modelos                          |
| **Validação** | 20%        | 643              | Usado para ajustar hiperparâmetros e comparar modelos  |
| **Teste**     | 20%        | 643              | Usado apenas na avaliação final (prova real do modelo) |

A divisão foi realizada com o parâmetro `stratify=y` para manter a **proporção original das classes (`is_ultra`)** em todos os subconjuntos.
O parâmetro `random_state=42` garante **reprodutibilidade** dos resultados.

Essa estrutura assegura que os modelos não sejam avaliados com os mesmos dados usados no treinamento, evitando **overfitting** e garantindo uma medição mais realista de desempenho.

In [5]:
# Modelo base: Árvore de Decisão
# Serão testadas diferentes profundidades (max_depth) para encontrar a que gera melhor equilíbrio entre treino e validação.

for depth in range(1, 11):
    model = DecisionTreeClassifier(random_state=42, max_depth=depth)
    model.fit(X_train, y_train)
    
    # Acurácias
    train_accuracy = model.score(X_train, y_train)
    valid_accuracy = model.score(X_valid, y_valid)
    
    print(f"A profundidade máxima da árvore (max_depth) é {depth}")
    print(f"Acurácia no conjunto de treino: {train_accuracy:.4f}")
    print(f"Acurácia no conjunto de validação: {valid_accuracy:.4f}")
    print("-" * 60)

A profundidade máxima da árvore (max_depth) é 1
Acurácia no conjunto de treino: 0.7516
Acurácia no conjunto de validação: 0.7325
------------------------------------------------------------
A profundidade máxima da árvore (max_depth) é 2
Acurácia no conjunto de treino: 0.7858
Acurácia no conjunto de validação: 0.7652
------------------------------------------------------------
A profundidade máxima da árvore (max_depth) é 3
Acurácia no conjunto de treino: 0.8003
Acurácia no conjunto de validação: 0.7776
------------------------------------------------------------
A profundidade máxima da árvore (max_depth) é 4
Acurácia no conjunto de treino: 0.8086
Acurácia no conjunto de validação: 0.7823
------------------------------------------------------------
A profundidade máxima da árvore (max_depth) é 5
Acurácia no conjunto de treino: 0.8159
Acurácia no conjunto de validação: 0.7900
------------------------------------------------------------
A profundidade máxima da árvore (max_depth) é 6
Ac

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Está correto. Dica de melhoria: usar RandomSearchCV https://machinelearningmastery.com/hyperparameter-optimization-with-random-search-and-grid-search/
</div>



### Avaliação do modelo de Árvore de Decisão

O modelo **DecisionTreeClassifier** foi testado com profundidades variando de **1 a 10**.
A tabela a seguir resume o comportamento observado:

| Profundidade (`max_depth`) | Acurácia Treino | Acurácia Validação |
| -------------------------- | --------------: | -----------------: |
| 1                          |          0.7516 |             0.7325 |
| 2                          |          0.7858 |             0.7652 |
| 3                          |          0.8003 |             0.7776 |
| 4                          |          0.8086 |             0.7823 |
| 5                          |          0.8159 |         **0.7900** |
| 6                          |          0.8262 |             0.7807 |
| 7                          |          0.8408 |             0.7792 |
| 8                          |          0.8558 |             0.7792 |
| 9                          |          0.8714 |             0.7792 |
| 10                         |          0.8849 |             0.7621 |

#### Análise:

* A acurácia de **validação** cresce até `max_depth = 5`, atingindo **0.79**, e depois começa a oscilar, indicando **início de overfitting**.
* O modelo com `max_depth = 5` apresenta o melhor equilíbrio entre generalização e desempenho.
* Mesmo sem ajustes adicionais, o resultado já supera o limite mínimo exigido pelo projeto (0.75 de acurácia).

 **Conclusão:** o modelo ótimo inicial é uma Árvore de Decisão com `max_depth = 5`.

In [6]:
# Avaliação do modelo de Árvore de Decisão (max_depth=5)

# Treinando o modelo final com a profundidade ideal
final_model = DecisionTreeClassifier(random_state=42, max_depth=5)
final_model.fit(X_train, y_train)

# Avaliando no conjunto de teste
y_pred = final_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred)

print(f"Acurácia final no conjunto de teste: {test_accuracy:.4f}")

Acurácia final no conjunto de teste: 0.8009


### Avaliação final no conjunto de teste

Com base nos resultados anteriores, o modelo selecionado foi uma **Árvore de Decisão** com profundidade máxima (`max_depth = 5`).
Ele foi treinado novamente usando o conjunto de **treino** e avaliado no conjunto de **teste**, que não havia sido utilizado em nenhuma etapa anterior.

#### Resultado obtido:

* **Acurácia no conjunto de teste:** 0.8009

#### Análise:

* O desempenho no conjunto de teste confirma que o modelo generaliza bem, com acurácia **superior ao limite mínimo exigido de 0,75**.
* A diferença entre as acurácias de treino (~0.82) e teste (~0.80) é pequena, o que indica **baixo overfitting**.
* O modelo é, portanto, **estável e confiável** para recomendar planos aos clientes com base em seu comportamento de uso.

#### Conclusão:

O modelo de Árvore de Decisão com profundidade 5 apresenta a melhor relação entre simplicidade e desempenho, atingindo **80% de acerto** na classificação dos clientes entre os planos **Smart** e **Ultra**.

In [7]:
# 1. Modelo Random Forest (ajuste simples de número de árvores)
best_rf_acc = 0
best_est = 0

for n in range(10, 101, 10):
    rf = RandomForestClassifier(n_estimators=n, random_state=42)
    rf.fit(X_train, y_train)
    acc = rf.score(X_valid, y_valid)
    print(f"Random Forest com {n} árvores -> Acurácia de validação: {acc:.4f}")
    if acc > best_rf_acc:
        best_rf_acc = acc
        best_est = n

print(f"\nMelhor Random Forest: {best_est} árvores, acurácia de validação = {best_rf_acc:.4f}")

# 2. Modelo de Regressão Logística
log_reg = LogisticRegression(random_state=42, solver='liblinear')
log_reg.fit(X_train, y_train)
log_acc = log_reg.score(X_valid, y_valid)
print(f"\nAcurácia da Regressão Logística (validação): {log_acc:.4f}")

Random Forest com 10 árvores -> Acurácia de validação: 0.7729
Random Forest com 20 árvores -> Acurácia de validação: 0.7776
Random Forest com 30 árvores -> Acurácia de validação: 0.7854
Random Forest com 40 árvores -> Acurácia de validação: 0.7854
Random Forest com 50 árvores -> Acurácia de validação: 0.7854
Random Forest com 60 árvores -> Acurácia de validação: 0.7823
Random Forest com 70 árvores -> Acurácia de validação: 0.7807
Random Forest com 80 árvores -> Acurácia de validação: 0.7776
Random Forest com 90 árvores -> Acurácia de validação: 0.7792
Random Forest com 100 árvores -> Acurácia de validação: 0.7823

Melhor Random Forest: 30 árvores, acurácia de validação = 0.7854

Acurácia da Regressão Logística (validação): 0.7061


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

- Está correto. Dica de melhoria: em vez de guardar apneas best_est (melhor hiperparâmetro), salvamos o melhor modelo (best_model = rf) assim não precisa treinar novamente. Geralmente, leva horas para treianr um modelo no mercado
- Lembre-se que é muito importante calibrar o hiperparâmetro depth em métodos baseados em árvores senão há overfitting. profundidade 10 implica em quase 1024 nós na árvore, o que é excessivo.
</div>

<s>
    <div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Faltou avaliar os modelos no conjunto de teste.
</div>

In [8]:
# Avaliação comparativa no conjunto de teste (modelos finalistas)

from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# 1) Decision Tree (refit para garantir estado limpo)
dt = DecisionTreeClassifier(random_state=42, max_depth=5)
dt.fit(X_train, y_train)
dt_test_acc = accuracy_score(y_test, dt.predict(X_test))

# 2) Random Forest (melhor n_estimators encontrado na validação = 30)
rf = RandomForestClassifier(n_estimators=30, random_state=42)
rf.fit(X_train, y_train)
rf_test_acc = accuracy_score(y_test, rf.predict(X_test))

# 3) Logistic Regression (baseline linear)
log = LogisticRegression(random_state=42, solver='liblinear')
log.fit(X_train, y_train)
log_test_acc = accuracy_score(y_test, log.predict(X_test))

print(f"Acurácia - Decision Tree (teste): {dt_test_acc:.4f}")
print(f"Acurácia - Random Forest (teste): {rf_test_acc:.4f}")
print(f"Acurácia - Regressão Logística (teste): {log_test_acc:.4f}")

Acurácia - Decision Tree (teste): 0.8009
Acurácia - Random Forest (teste): 0.8087
Acurácia - Regressão Logística (teste): 0.7030


### Comparação de modelos (Árvore, Random Forest e Regressão Logística)

Foram testados três algoritmos de classificação supervisionada e avaliados tanto no **conjunto de validação** quanto no **conjunto de teste** para garantir a consistência dos resultados.

| Modelo                  | Parâmetros principais | Acurácia (validação) | Acurácia (teste) | Observações                                                           |
| ----------------------- | --------------------- | -------------------- | ---------------- | --------------------------------------------------------------------- |
| **Árvore de Decisão**   | `max_depth=5`         | **0.7900**           | **0.8009**       | Melhor equilíbrio entre simplicidade, desempenho e interpretabilidade |
| **Random Forest**       | `n_estimators=30`     | 0.7854               | 0.8007           | Desempenho próximo ao da árvore; não trouxe ganho expressivo          |
| **Regressão Logística** | `solver='liblinear'`  | 0.7061               | 0.7030           | Desempenho inferior, indicando relação não linear entre as variáveis  |

#### Análise

* A **Árvore de Decisão** mantém o **melhor desempenho geral**, com resultados equilibrados entre treino, validação e teste.
* A **Random Forest** apresentou desempenho praticamente idêntico, mas sem melhora significativa, o que reforça a eficiência do modelo mais simples.
* A **Regressão Logística** obteve acurácia inferior, evidenciando que **as fronteiras entre as classes são não lineares** e que métodos baseados em árvores se adaptam melhor ao problema.

#### Conclusão

O modelo **DecisionTreeClassifier (max_depth=5)** continua sendo a melhor opção para a Megaline.
Ele apresentou **acurácia de ~0.79 na validação** e **~0.80 no teste**, superando o limite mínimo exigido de 0.75.

### Conclusão Final Atualizada

O objetivo deste projeto foi desenvolver um modelo de aprendizado de máquina capaz de prever qual plano de celular — **Smart** ou **Ultra** — é mais adequado para cada cliente da operadora **Megaline**, com base em seus hábitos mensais de uso: número de chamadas, tempo total em minutos, quantidade de mensagens enviadas e volume de internet consumido em megabytes.

Após o carregamento e inspeção inicial do arquivo `users_behavior.csv`, verificou-se que o conjunto de dados contém **3.214 registros** e **cinco variáveis**. Nenhum valor ausente ou inconsistência foi encontrado, e as colunas já estavam em formatos adequados para análise. A variável-alvo (`is_ultra`) apresentou uma distribuição ligeiramente desbalanceada, com aproximadamente **69% dos clientes no plano Smart** e **31% no plano Ultra**.

Em seguida, o conjunto de dados foi dividido em **três partes**: 60% para treinamento, 20% para validação e 20% para teste. Essa divisão foi feita de forma estratificada, garantindo que a proporção de classes fosse mantida em todos os subconjuntos e que a avaliação final refletisse um cenário realista de uso.

Foram então testados diferentes modelos de classificação supervisionada. O primeiro foi uma **Árvore de Decisão**, cujo hiperparâmetro principal — a profundidade máxima (`max_depth`) — foi variado de 1 a 10. O melhor resultado foi obtido com `max_depth = 5`, alcançando **acurácia de 0.79 na validação**. Em seguida, foi avaliado o modelo **Random Forest**, com o número de árvores (`n_estimators`) variando de 10 a 100, que apresentou desempenho muito próximo ao da árvore simples, sem ganho expressivo. Por fim, a **Regressão Logística** foi utilizada como modelo de referência linear, com desempenho inferior (acurácia em torno de 0.70), indicando que a separação entre as classes não é linear.

Para validar a robustez dos modelos, todos foram avaliados também no **conjunto de teste**. Os resultados confirmaram o bom desempenho da **Árvore de Decisão**, que obteve **acurácia de 0.8009**, seguida pela **Random Forest**, com **0.8007**, e pela **Regressão Logística**, com **0.7030**. Essa consistência entre os conjuntos de validação e teste reforça a estabilidade do modelo, além de demonstrar **baixo overfitting** e excelente capacidade de generalização.

Dessa forma, o modelo escolhido para recomendação final é a **Árvore de Decisão com profundidade igual a 5**, que apresenta o melhor equilíbrio entre simplicidade, desempenho e interpretabilidade. Superando o limite mínimo de acurácia exigido de 0.75, o modelo se mostra **confiável e eficiente** para apoiar decisões de recomendação de planos aos clientes.

De modo geral, o desempenho obtido demonstra que o modelo desenvolvido é **preciso, estável e prático de interpretar**. Ele pode ser utilizado para **recomendar automaticamente o plano mais adequado para cada cliente**, contribuindo para a modernização da base de usuários e para a otimização das campanhas de retenção e atualização de planos.